src/Uno.UI/UI/Xaml/Controls/TextBlock/TextVisual.skia.cs

```diff
@@ -27,7 +27,7 @@ public TextVisual(Compositor compositor, TextBlock owner) : base(compositor)
 			_owner = new WeakReference<TextBlock>(owner);
 		}
 
-		internal override void Draw(in DrawingSession session)
+		internal override void Draw(in DrawingSession session, in SKMatrix initialTransform)
```

Comment:
The `TotalMatrix` does not take into account the initial state of the canvas (e.g. scaling or offsets), only the transformations that are done by the UIElements/Visuals. So, we need to apply the "initial state of the canvas" before we apply the `TotalMatrix`.

Previously, we modified (i.e. left multiplied) the canvas matrix when we `BeginDrawing` each visual, so the effect on the root would propagate everywhere. Now, we need to reapply it whenever we set the matrix using `TotalMatrix`.

This is only relevant to `RenderTargetBitmap` and high DPI renderers. the `initialTransform` will be the identity matrix otherwise.

Reply:
> Now, we need to reapply it whenever we set the matrix using TotalMatrix.

Not sure to understand why? The `TotalMatrix` is "the transform of the visual relative to the root of the visual tree", if the root is already transformed (e.g. DPI) it should be applied when we start the drawing but should not impact the `TotalMatrix`?

> This is only relevant to RenderTargetBitmap and high DPI renderers. the initialTransform will be the identity matrix otherwise.

That's what I was suspecting, in that case why not pre-concat parent's `TotalMatrix.Inverse()` before drawing the content?

Reply:
> if the root is already transformed (e.g. DPI) it should be applied when we start the drawing but should not impact the TotalMatrix?

Exactly.

When we draw a visual, we directly `SKCanvas.SetMatrix` the precalculated `TotalMatrix`, so any initial matrix set on the SKCanvas will be lost. To counteract this, we pass this initial matrix along so that when we copy (not multiply) the TotalMatrix to the canvas matrix, we take into account this initial matrix.

Alternatively, we could rely on the fact that the parent was drawn before the current visual was drawn, so instead of SetMatrix, we could just preconcatenate the transformation from `current.TotalMatrix` to `parent.TotalMatrix` (i.e. `current.TotalMatrix * parent.TotalMatrix.Inverse()`). This indeed was the initial implementation, but this fails with `RenderTargetBitmap`, where the parent might not be part of the drawing.

Reply:
Ok get it, it's because I was still under the original idea to capture the `TotalMatrix` provided by the canvas itself during drawing, but finally it's the kind-off the opposite, we are computing it and forcefully setting it on the canvas by our own.

> we could rely on the fact that the parent was drawn before the current visual was drawn

That's how we work currently ... and if we are no longer relying on this, we can remove all the `Canvas.Save` and `Canvas.Restore`. Not sure yet what is the best approach ...

> but this fails with RenderTargetBitmap, where the parent might not be part of the drawing.

This case can be handled in the `RenderRootVisual` (which is already the case with the `ignoreLocation` flag ... where we should actually use the `Parent.TotalTransform.Inverse()` instead of only translating to the opposite offsets).



In all cases, the `initialTransform` should be moved into the `DrawingSession` and IMO should be named something like `OriginTransform` or `RootTransform` as it's only about getting the origin of the drawing plan.

Reply:
The problem with "caching" the canvas' TotalMatrix during drawing is that it will give outdated values if you modify an element and request TransformToVisual before we've had a chance to redraw.

In fact, this is exactly what the original version did. It would fall back to the old `GetTransform` implementation until we redraw, but it's cleaner to do it in reverse: calculate the TotalMatrix on the spot whenever you need it, and don't recalculate it on redrawing.

About removing the `Save`/`Restore` calls, I guess we might still need them for the clipping/filters. I'm kind of too scared of possible edge cases in some weird visuals :)